tools/chpldoc/findUndocumentedSymbols.py

```diff
@@ -0,0 +1,387 @@
+                node.parent(), (dyno.Module, dyno.AggregateDecl)
+            ),
+        ),
+        "enum_element": (dyno.EnumElement, None),
+        "interface": (
+            dyno.Interface,
+            lambda node, match: isinstance(
+                node.parent(), (dyno.Module, dyno.AggregateDecl)
+            ),
+        ),
+    }
+
+    def _get_previous_sibling(self, node: dyno.AstNode):
+        parent = node.parent()
+        if not parent:
+            for sib1, sib2 in lookAhead(self.ast):
+                if sib2 and node.unique_id() == sib2.unique_id():
+                    return sib1
+            return None
+
+        for sib1, sib2 in lookAhead(parent):
+            if sib2 and node.unique_id() == sib2.unique_id():
+                return sib1
+        return None
+
+    def get_documentable_symbols(self, root: dyno.AstNode) -> Generator:
+        for (
+            pat,
+            check_func,
+        ) in FindUndocumentedSymbols.documentable_symbol_patterns.values():
+            matches = [
+                m for m in each_matching(root, pat) if not check_func or check_func(*m)
+            ]
+            for node, _ in matches:
+                yield node
+
+    def has_doc_comment(self, node: dyno.AstNode) -> bool:
+        """
+        This node has a doc comment if
+        - the previous node is a comment
+        """
+
+        if (
+            (prev := self._get_previous_sibling(node))
+            and isinstance(prev, dyno.Comment)
+            and is_docstring_comment(prev)
+        ):
+            return True
+
+        return False
+
+    def __call__(self):
+        syms = [s for s in self.get_documentable_symbols([a for a in self.ast])]
+        for s in syms:
+            if not is_nodoc(s) and not self.has_doc_comment(s):
+                if self.ignore_deprecated and is_deprecated(s):
+                    continue
+                if self.ignore_unstable and is_unstable(s):
+                    continue
+                yield s
+
+
+def get_files(files: List[str], check_suffix=False) -> Generator:
+    """
+    Yield all chapel files, following directories recursively
+    """
+
+    for file in files:
+        if os.path.isfile(file):
+            # just yield any file given if not checking the suffix
+            if not check_suffix:
+                yield file
+            elif check_suffix and os.path.splitext(file)[1] == ".chpl":
+                yield file
+
+        elif os.path.isdir(file):
+            for root, subdirs, files in os.walk(file):
+                # yield files in directory files, always check the suffix
+                yield from get_files([os.path.join(root, file) for file in files], True)
+                # recurse into sub dirs, always checl the suffix
+                yield from get_files([os.path.join(root, d) for d in subdirs], True)
+        else:
+            print("Error: path was not file or directory", file=sys.stderr)
+            exit(1)
```

Comment:
Can a path glob not be used for this purpose?

https://docs.python.org/3/library/glob.html#glob.glob

I'd imagine that if a path doesn't end in *.chpl, a pattern like `dir/**/*.chpl` would suffice.

Reply:
d'oh!